### DISCUSSION TITLE
Proposal: rename `assertion` to `test`

### DISCUSSION THREAD

**Discussion by fantix** [moonbitlang/core#142] [General]
## Motivations

* `assert_*()` no longer panics, returning a test-specific `Result[Unit, String]` now.

   I remember the `@assertion` package was added to the stdlib before tests were switched to return Results. This indicates the `assert_*()` functions are not "assertions" that don't get in the way of non-test user code. Now the users have to handle the Results properly in production code, especially to deal with the `String` errors.

   With that said, a "proper" assertion would have to be like:

   ```
   @assertion.assert_eq(1, 1).unwrap()
   ```

   However, this explicitness of `unwrap()` defeats the idea of calling the package `@assertion`.

* There is no way to turn off assertions.

   This is debatable - Java and Python provide options to turn off assertions at runtime, C/C++ could use some macro tricks, and Rust disallows turning off `std::assert` calls but provides `std::debug_assert` that is excluded in release builds. The `@assertion` package in MoonBit looks more like a testing library to me, especially since it covers `eq`, `ne`, `true`, `false`, and likely more to add.

   In addition, we could borrow `debug_assert` from Rust that 1) panics if evaluated false, 2) is excluded from release builds, 3) does not live in the `@assertion` (or `@test`) package, and 4) has no siblings like `eq`, `ne` and so on.

* `@assertion.assert...` is redundant to type.

   Typing `@test.assert_eq()` feels much more natural than repeating the "assert" word twice.

**Comment by Yoorkin** [comment]
```
@test.assert_eq(a,b)?
@test.assert_is_not(a,b)?
```
or
```
@test.eq(a,b)?
@test.is_not(a,b)?
```
looks good to me, but the `test` is a keyword of MoonBit. 
So I would prefer to use another package name to replace `assertion`, if a suitable one exists.

**Comment by zxch3n** [comment]
You may use https://github.com/zxch3n/moon-expect

# Example Usage

```
expect("hello").equal("hello")
expect("hello").not().equal("hello") // abort
expect("hello").equal("hello1") // abort
expect("hello").not().equal("hello1") 

let a : Option[Int] = Some(1)
let b : Option[Int] = None
expect(a).to_be_some() 
expect(b).to_be_none() 

expect("hello").to_have_length(5)
expect("hello").not().to_have_length(6)

expect(1).to_be_less_than(4)

let a : Result[Int, Int] = Ok(1)
let b : Result[Int, Int] = Err(1)
expect(a).to_be_ok()
expect(b).to_be_err()

expect(1.0).close_to(1.0, 0.1)
```

## Example Error Messages

```
AssertionError: expected 1.0 to be close to 1.2 (±0.1)

- Expected: 
+ Received: 

- 1.2
+ 1.0
```

```
AssertionError: expected "hello" to equal "hello1"

- Expected: 
+ Received: 

- "hello1"
+ "hello"
```

I'd love to make it support something like `debug_assert` if possible.

**Comment by bobzhang** [comment]
Note test is a keyword in MoonBit, but `@test.eq` also works, the name `test` does look better.

Any other thoughts?

**Comment by Yoorkin** [comment]
The package `@assertion` has been renamed to `@test`. The names of the testing functions have also changed to `eq` `ne` ... `is_true`.   #609
